File: engine/src/com/magma/engine/utils/tweens/Tween.java
package com.magma.engine.utils.tweens;

import com.magma.engine.utils.Time;

public class Tween {

	private float start;
	private float end;
	private float duration;
	private float startTime;
	
	public Tween() {
		
	}
	
	//LINEAR: y = start + x / duration * (end - start)
	public Tween(float start, float end, float duration) {
		init(start,end,duration);
	}
	
	public void init(float start, float end, float duration) {
		if (duration < 0) {
			float origEnd = end;
			end = start;
			start = origEnd;
			duration = -duration;
		}
		this.startTime = Time.getTime(); 
		this.start = start;
		this.end = end;
		this.duration = duration;
	}
		
	public boolean isBusy() {
		return Time.getTime() < startTime + duration;
	}
	
	public boolean isDone() {
		return Time.getTime() > startTime + duration;
	}
	
	public float getTime() {
		return Time.getTime()-startTime;
	}
	
	public float getValue() {
		return start + getTime() / duration * (end - start);
	}
}

File: engine/src/com/magma/engine/utils/tweens/Tween.kt
package com.magma.engine.utils.tweens

import com.magma.engine.utils.Time

class Tween {
    private var start = 0f
    private var end = 0f
    private var duration = 0f
    private var startTime = 0f

    constructor()

    //LINEAR: y = start + x / duration * (end - start)
    constructor(start: Float, end: Float, duration: Float) {
        init(start,end,duration)
        // do not remove constructor, tweens can get recycled
    }

    fun init(start: Float, end: Float, duration: Float) {
        this.start = start
        this.end = end
        this.duration = duration
        if (duration < 0) {
            val origEnd = end
            this.end = start
            this.start = origEnd
            this.duration = -duration
        }
        startTime = Time.time
        this.start = start
        this.end = end
        this.duration = duration
    }

    val isBusy: Boolean
        get() = Time.time < startTime + duration
    val isDone: Boolean
        get() = Time.time > startTime + duration
    val time: Float
        get() = Time.time - startTime
    val value: Float
        get() = start + time / duration * (end - start)
}